suppressMessages(library("ddpcr"))
suppressMessages(library("optparse"))
#
packages <- c("arm","biomaRt","rtracklayer","GenomicRanges","GenomicFeatures", "dplyr", "reshape2", "data.table","tidyverse",
               "readxl","parallel","qvalue","Matrix","ggplot2",
               "MASS","stringr","fastglm","pscl")
#
invisible(quiet(lapply(packages, library, character.only = TRUE), all = T)) # invisible for suppressing output of the lapply, suppresMessages for libraries



#' Compute DiffInVEx Regression Coefficients
#'
#' This function computes DiffInVEx regression coefficients for a list of genes. The DiffInVEx approach is used to 
#' analyze the impact of genomic variations on gene expression. The function processes genomic tracks, mutation 
#' data, and other relevant variables to compute the coefficients, which are then saved to the specified output 
#' directory. Parallel processing is supported to speed up computation.
#'
#' @param gene_list Character vector. A list of gene symbols for which DiffInVEx coefficients are to be computed.
#' @param mutation_file Character string. File path to the mutation data file in a format compatible with the 
#'        `MkGRanges_SNPs` function.
#' @param annotation_file Character string. File path to the sample annotation file.
#' @param variable_file Character string. File path to the file containing variables to control for in the regression.
#' @param tool_directory Character string. Directory containing auxiliary R scripts and tool-related files.
#' @param output_directory Character string. Directory where the results will be saved.
#' @param no_cores Integer. Number of cores to use for parallel processing. Default is 1.
#' @param reference_genome Character string. Reference genome to use. Default is "hg19".
#' @param DiffInVEx_BW Integer. Background width for DiffInVEx analysis. Default is 50.
#' @param DiffInVEx_cluster Integer. Number of clusters for DiffInVEx analysis. Default is 0.
#' @param DiffInVEx_mode Integer. Mode of DiffInVEx analysis. Default is 1.
#' @param sample_annotation_perGene_file Character string. File path to the per-gene sample annotation file, if applicable.
#'
#' @details
#' The function begins by loading the necessary genomic tracks and mutation data. If the DiffInVEx mode is set to 4, 
#' it skips the loading of these tracks. It then processes the mutation data and sample annotations, if applicable, 
#' and reads genomic features if required by the DiffInVEx mode. The function uses parallel processing to compute the 
#' DiffInVEx selection coefficients for each gene in the list. The results are saved to the specified output directory.
#'
#' @return NULL. Results are saved to the specified output directory in the form of files generated by the 
#'         `get_InVEx_selection_coefficients` function.
#'
#' @importFrom parallel mclapply
#'
#' @export
diffinvex_coefficients <- function(gene_list,
                                   mutation_file,
                                   annotation_file,
                                   variable_file,
                                   tool_directory,
                                   output_directory,
                                   no_cores= 1,
                                   reference_genome = "hg19",
                                   DiffInVEx_BW = 50,
                                   DiffInVEx_cluster = 0,
                                   DiffInVEx_mode = 1,
                                   sample_annotation_perGene_file = NULL){
        #--------------------------#
        # target/background tracks #
        #--------------------------#
        print("Data loading ...")
        start_time = Sys.time()
        #
        if(DiffInVEx_mode != 4){
            # use pre-computed target and background tracks
            trBg_tracks <- load_tracks(tool_directory,reference_genome,DiffInVEx_BW,DiffInVEx_cluster)
            trTracks <- trBg_tracks[["target"]]
            bgTracks <- trBg_tracks[["background"]]
            gTracks  <- trBg_tracks[["gene"]]
          } else{
            trTracks <- NULL
            bgTracks <- NULL
            gTracks  <- NULL
          }
        #
        #------#
        #-SNVs-#
        #------#
        if(DiffInVEx_mode != 4){
            sample_annotation  <- load_sample_annotation(annotation_genome_wide_file = annotation_file)
            mutationData       <- MkGRanges_SNPs(mutation_file, gTracks, sample_annotation)
            sample_annotation  <- add_MutationLoad(mutationData,sample_annotation,variable_file,trTracks,bgTracks)
        } else{
            sample_annotation <- NULL
            mutationData <- NULL
        }
        #
        #----------------#
        #-Genomic Tracks-#
        #----------------#
        filtering_tracks <- NULL
        filter_inclusion <- TRUE
        #
        # read genomic tracks if exonic/intronic regions will be computed
        if(DiffInVEx_mode != 1){
            quiet(source(paste0(tool_directory,"/R/GettingGenomicFeatures.R")), all = T)
            genomic_tracks <- read_genomic_tracks(reference_genome, tool_directory, filtering_tracks, filter_inclusion)
        } else{
            genomic_tracks <- NULL
        }
        #
        end_time = Sys.time()
        end_time - start_time
        #
        #############
        # variables #
        #############
        opt <- list()
        opt$mutations_data                     <- mutationData
        opt$sample_annotation                  <- sample_annotation
        opt$variables_to_control_file          <- variable_file
        opt$sample_annotation_perGene_file     <- sample_annotation_perGene_file
        opt$genomic_tracks                     <- genomic_tracks
        #
        opt$tool_directory   <- tool_directory
        opt$output_directory <- output_directory
        #
        opt$cluster_number   <- DiffInVEx_cluster
        opt$reference_genome <- reference_genome
        opt$diffInVEx_mode   <- DiffInVEx_mode
        opt$diffInVEx_BW     <- DiffInVEx_BW
        #
        if (!dir.exists(opt$output_directory)){
          dir.create(opt$output_directory)
        }
        #
        ######################################
        ### compute selection coefficients ###
        ######################################
        print("processing ...")
        start_time = Sys.time()
        ##
        ##-Parallel code-##
        #
        mclapply(gene_list, 
                 function(gene) get_InVEx_selection_coefficients(HGNC_symbol = gene,
                                                                clusterNumber = opt$cluster_number,
                                                                mutations = opt$mutations_data, 
                                                                sample_annotation = opt$sample_annotation, 
                                                                sample_annotation_perGene_file = opt$sample_annotation_perGene_file, 
                                                                controlled_variables_input = opt$variables_to_control_file,
                                                                genomic_tracks = opt$genomic_tracks,  
                                                                outputDirectory = opt$output_directory,
                                                                toolDirectory = opt$tool_directory,
                                                                refGenome  = opt$reference_genome,
                                                                diffInVEx_mode = opt$diffInVEx_mode,
                                                                diffInVEx_BW = opt$diffInVEx_BW),
                  mc.cores = no_cores)
        #
        end_time = Sys.time()
        end_time - start_time
}